### PR TITLE
feat(provider/cf): add SG clone action

### PR DESCRIPTION
**Pull request by jmelchio**
Co-Authored-By: Joris Melchior <joris.melchior@gmail.com>
Co-Authored-By: Stu Pollock <spollock@pivotal.io>

spinnaker/spinnaker#3656

### CODE REVIEW

**Comment by jmelchio** [comment]
Needs to go in sync with: https://github.com/spinnaker/orca/pull/2551

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java [hunk 1]
```diff
@@ -85,7 +85,7 @@ public DeployCloudFoundryServerGroupDescription convertDescription(Map input) {
     converted.setSpace(findSpace(converted.getRegion(), converted.getClient())
       .orElseThrow(() -> new IllegalArgumentException("Unable to find space '" + converted.getRegion() + "'.")));
 
-    Map artifactSource = (Map) input.get("artifactSource");
+    Map artifactSource = (Map) input.get("artifact");
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java]: Can you elaborate on why the property name changed?

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/CloneCloudFoundryServerGroupAtomicOperationConverter.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/*
```
Comment (jkschneider) [inline clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/CloneCloudFoundryServerGroupAtomicOperationConverter.java]: Why is a separate operation necessary if it extends from and doesn't modify the behavior of `DeployCloudFoundryServerGroupAtomicOperationConverter`?

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java (hunk 1 above)
Comment (jmelchio) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java]: The mapping from `artifact` to `artifactSource` happens in Orca. For the create server group operation we have control over that so it works fine, when sending the clone operation we don't have an easy extension point in Orca so we considered it easier to adjust this than to try and implement a custom operation for clone in Orca.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/CloneCloudFoundryServerGroupAtomicOperationConverter.java (hunk 2 above)
Comment (jmelchio) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/CloneCloudFoundryServerGroupAtomicOperationConverter.java]: Core Deck determines that you execute a clone action and puts that operation string on it. As a result you have to have a handler for it on the receiving end in clouddriver.

File: clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java (hunk 1 above)
Comment (jkschneider) [reply clouddriver-cloudfoundry/src/main/java/com/netflix/spinnaker/clouddriver/cloudfoundry/deploy/converters/DeployCloudFoundryServerGroupAtomicOperationConverter.java]: Thanks @jmelchio!

**Pull request merged**